ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+      ptr->semid = -1;
+    } else {
+      rb_warn("semctl: failed to remove semaphore with semid %d, errno %d (%s)",ptr->semid, errno, strerror(errno));
+    }
+  } else {
+    //rb_warn("semctl: semaphore set with semid %d deleted", ptr->semid);
+    ptr->semid = -1;
+  }
+  return self;
+}
+
+
+
+
+/*
+ * semian_cb_data_lock/unlock and associated functions decrement/increment semaphore
+ */
+
+static VALUE
+semian_cb_data_lock(VALUE self)
+{
+  return (VALUE) WITHOUT_GVL(semian_cb_data_lock_without_gvl, (void *)self, RUBY_UBF_IO, NULL);
+}
+
+static void *
+semian_cb_data_lock_without_gvl(void *self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct((VALUE)self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (ptr->lock_triggered)
+    return (void *)Qtrue;
+  if (-1 == ptr->semid){
+    rb_raise(eInternal, "semid not set, errno %d: (%s)", errno, strerror(errno));
+    return (void *)Qfalse;
+  }
+  VALUE retval;
+
+  struct timespec ts = { 0 };
+  ts.tv_sec = kCBInternalTimeout;
+
+  if (-1 == semtimedop(ptr->semid,&decrement,1, &ts)) {
+    rb_raise(eInternal, "error with semop locking,  %d: (%s)", errno, strerror(errno));
+    retval=Qfalse;
+  } else
+    retval=Qtrue;
+
+  ptr->lock_triggered = true;
+  //rb_warn("semop: lock success");
+  return (void *)retval;
+}
+
+static VALUE
+semian_cb_data_unlock(VALUE self)
+{
+  return (VALUE) WITHOUT_GVL(semian_cb_data_unlock_without_gvl, (void *)self, RUBY_UBF_IO, NULL);
+}
+
+static void *
+semian_cb_data_unlock_without_gvl(void *self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct((VALUE)self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (!(ptr->lock_triggered))
+    return (void *)Qtrue;
+  if (-1 == ptr->semid){
+    rb_raise(eInternal, "semid not set, errno %d: (%s)", errno, strerror(errno));
+    return (void *)Qfalse;
+  }
+  VALUE retval;
+
+  struct timespec ts = { 0 };
+  ts.tv_sec = kCBInternalTimeout;
+
+  if (-1 == semtimedop(ptr->semid,&increment,1 , &ts)) {
+    rb_raise(eInternal, "error with semop unlocking, errno: %d (%s)", errno, strerror(errno));
+    retval=Qfalse;
+  } else
+    retval=Qtrue;
+
+  ptr->lock_triggered = false;
+  //rb_warn("semop unlock success");
+  return (void *)retval;
+}
+
+
+
+
+/*
+  Acquire memory by getting shmid, and then attaching it to a memory location,
+    requires semaphore for locking/unlocking to be setup
+*/
+static VALUE
+semian_cb_data_acquire_memory(VALUE self, VALUE permissions)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+
+  if (-1 == ptr->semid){
+    rb_raise(eInternal, "semid not set, errno %d: (%s)", errno, strerror(errno));
+    return self;
+  }
+  if (TYPE(permissions) != T_FIXNUM)
+    rb_raise(rb_eTypeError, "expected integer for permissions");
+
+  if (!semian_cb_data_lock(self))
+    return self;
+
+  key_t key = ptr->key;
+  if (-1 == (ptr->shmid = shmget( key,
```

Comment:
Throwing this out there - but it might be better to use POSIX `shm_open` and then `mmap` the file here instead of `shmget`. There's one big advantage: we can use a full filename instead of a hash key.